// @dimforge/rapier2d/pipeline/event_queue.js
import { RawEventQueue } from '../raw';
/// Flags indicating what events are enabled for colliders.
export var ActiveEvents;
(function (ActiveEvents) {
    /// Enable intersection events.
    ActiveEvents[ActiveEvents["INTERSECTION_EVENTS"] = 1] = "INTERSECTION_EVENTS";
    /// Enable contact events.
    ActiveEvents[ActiveEvents["CONTACT_EVENTS"] = 2] = "CONTACT_EVENTS";
})(ActiveEvents || (ActiveEvents = {}));
/**
 * A structure responsible for collecting events generated
 * by the physics engine.
 *
 * To avoid leaking WASM resources, this MUST be freed manually with `eventQueue.free()`
 * once you are done using it.
 */
var EventQueue = /** @class */ (function () {
    /**
     * Creates a new event collector.
     *
     * @param autoDrain -setting this to `true` is strongly recommended. If true, the collector will
     * be automatically drained before each `world.step(collector)`. If false, the collector will
     * keep all events in memory unless it is manually drained/cleared; this may lead to unbounded use of
     * RAM if no drain is performed.
     */
    function EventQueue(autoDrain, raw) {
        this.raw = raw || new RawEventQueue(autoDrain);
    }
    /**
     * Release the WASM memory occupied by this event-queue.
     */
    EventQueue.prototype.free = function () {
        this.raw.free();
        this.raw = undefined;
    };
    /**
     * Applies the given javascript closure on each contact event of this collector, then clear
     * the internal contact event buffer.
     *
     * @param f - JavaScript closure applied to each contact event. The
     * closure should take three arguments: two integers representing the handles of the colliders
     * involved in the contact, and a boolean indicating if the contact started (true) or stopped
     * (false).
     */
    EventQueue.prototype.drainContactEvents = function (f) {
        this.raw.drainContactEvents(f);
    };
    /**
     * Applies the given javascript closure on each intersection event of this collector, then clear
     * the internal intersection event buffer.
     *
     * @param f - JavaScript closure applied to each intersection event. The
     * closure should take four arguments: two integers representing the handles of the colliders
     * involved in the intersection, and a boolean indicating if they started intersecting (true) or
     * stopped intersecting (false).
     */
    EventQueue.prototype.drainIntersectionEvents = function (f) {
        this.raw.drainIntersectionEvents(f);
    };
    /**
     * Removes all events contained by this collector
     */
    EventQueue.prototype.clear = function () {
        this.raw.clear();
    };
    return EventQueue;
}());
export { EventQueue };
